File: ios/Sources/InstallReferrerPlugin/InstallReferrerPlugin.swift
import Foundation
import Capacitor

@objc(InstallReferrerPlugin)
public class InstallReferrerPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "InstallReferrerPlugin"
    public let jsName = "InstallReferrer"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "getReferrer", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "GetReferrer", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getPluginVersion", returnType: CAPPluginReturnPromise)
    ]

    private let implementation = InstallReferrer()

    @objc func getReferrer(_ call: CAPPluginCall) {
        resolveReferrer(call)
    }

    @objc func GetReferrer(_ call: CAPPluginCall) {
        resolveReferrer(call)
    }

    @objc func getPluginVersion(_ call: CAPPluginCall) {
        call.resolve([
            "version": implementation.getPluginVersion()
        ])
    }

    private func resolveReferrer(_ call: CAPPluginCall) {
        do {
            let token = try implementation.getAttributionToken()
            let fetchAppleAttribution = call.getBool("fetchAppleAttribution") ?? false
            var result: [String: Any] = [
                "platform": "ios",
                "attributionToken": token
            ]

            guard fetchAppleAttribution else {
                call.resolve(result)
                return
            }

            let retryCount = call.getInt("appleAttributionRetryCount") ?? 3
            let retryDelayMs = call.getInt("appleAttributionRetryDelayMs") ?? 5_000

            Task {
                do {
                    let attribution = try await implementation.fetchAppleAttribution(
                        token: token,
                        retryCount: retryCount,
                        retryDelayMs: retryDelayMs
                    )
                    result["appleAttribution"] = attribution
                    call.resolve(result)
                } catch {
                    reject(call, error: error)
                }
            }
        } catch {
            reject(call, error: error)
        }
    }

    private func reject(_ call: CAPPluginCall, error: Error) {
        if let installReferrerError = error as? InstallReferrerError {
            call.reject(installReferrerError.message, installReferrerError.code, error)
            return
        }

        call.reject(error.localizedDescription, "IOS_ATTRIBUTION_ERROR", error)
    }
}

File: android/src/main/java/app/capgo/installreferrer/InstallReferrerPlugin.java
package app.capgo.installreferrer;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

@CapacitorPlugin(name = "InstallReferrer")
public class InstallReferrerPlugin extends Plugin {

    private final InstallReferrer implementation = new InstallReferrer();

    @PluginMethod
    public void getReferrer(PluginCall call) {
        implementation.getReferrer(
            getContext(),
            new InstallReferrer.ReferrerCallback() {
                @Override
                public void onSuccess(JSObject result) {
                    call.resolve(result);
                }

                @Override
                public void onError(String code, String message, Exception error) {
                    Exception resolvedError = error != null ? error : new Exception(message);
                    call.reject(message, code, resolvedError);
                }
            }
        );
    }

    @PluginMethod
    public void GetReferrer(PluginCall call) {
        getReferrer(call);
    }

    @PluginMethod
    public void getPluginVersion(PluginCall call) {
        JSObject ret = new JSObject();
        ret.put("version", implementation.getPluginVersion());
        call.resolve(ret);
    }
}
